File: src/connectors.ts
import type { Corti } from "@corti/sdk";
import type {
  A2aConnector,
  ConnectorDef,
  CortiAgentConnector,
  McpConnector,
  RegistryConnector,
} from "./types";

// ── Connector factory helpers ────────────────────────────────────────────────

/**
 * Pre-built connector factories for common patterns.
 *
 * @example
 * ```ts
 * connectors.fromAgent({ agentId: subAgent.id })
 * connectors.mcp({ mcpUrl: "https://mcp.corti.ai" })
 * connectors.registry({ name: "@corti/medical-coding" })
 * ```
 */
export const connectors = {
  /** Reference another Corti agent as a sub-agent connector. */
  fromAgent: (opts: { agentId: string }): CortiAgentConnector => ({
    type: "cortiAgent",
    agentId: opts.agentId,
  }),

  /** Attach an MCP server directly to the agent. */
  mcp: (opts: {
    mcpUrl: string;
    name?: string;
    transport?: "sse" | "streamable_http";
    token?: string;
  }): McpConnector => ({
    type: "mcp",
    mcpUrl: opts.mcpUrl,
    ...(opts.name !== undefined && { name: opts.name }),
    ...(opts.transport !== undefined && { transport: opts.transport }),
    ...(opts.token !== undefined && { token: opts.token }),
  }),

  /** Reference a named expert from the Corti registry. */
  registry: (opts: { name: string; systemPrompt?: string }): RegistryConnector => ({
    type: "registry",
    name: opts.name,
    ...(opts.systemPrompt !== undefined && { systemPrompt: opts.systemPrompt }),
  }),

  /** Connect via the A2A protocol (not yet supported — reserved for future use). */
  a2a: (opts: { a2aUrl: string }): A2aConnector => ({
    type: "a2a",
    a2aUrl: opts.a2aUrl,
  }),
};

// ── Internal: map ConnectorDef[] → AgentsCreateAgentExpertsItem[] ────────────

function mcpUrlToName(url: string): string {
  const match = url.match(/^https?:\/\/([^/?#]+)/);
  const hostname = match?.[1] ?? "";
  return (
    hostname
      .replace(/\./g, "-")
      .replace(/[^a-z0-9-]/gi, "")
      .slice(0, 48) || "mcp-server"
  );
}

export function connectorsToExperts(
  defs: ConnectorDef[]
): Corti.AgentsCreateAgentExpertsItem[] {
  return defs.map((conn): Corti.AgentsCreateAgentExpertsItem => {
    switch (conn.type) {
      case "mcp": {
        const name = conn.name ?? mcpUrlToName(conn.mcpUrl);
        return {
          type: "new",
          name,
          description: `MCP server at ${conn.mcpUrl}`,
          mcpServers: [
            {
              name,
              transportType: conn.transport ?? "sse",
              authorizationType: conn.token ? "bearer" : "inherit",
              url: conn.mcpUrl,
              ...(conn.token !== undefined && { token: conn.token }),
            },
          ],
        };
      }
      case "registry":
        return {
          type: "reference",
          name: conn.name,
          ...(conn.systemPrompt !== undefined && { systemPrompt: conn.systemPrompt }),
        };
      case "cortiAgent":
        return {
          type: "reference",
          id: conn.agentId,
        };
      case "a2a":
        throw new Error(
          `[AgentSDK] A2A connectors are not yet supported (url: ${conn.a2aUrl}). ` +
            `Use type "mcp" with an MCP-compatible endpoint instead.`
        );
      default: {
        const exhaustive: never = conn;
        throw new Error(`[AgentSDK] Unknown connector type: ${(exhaustive as ConnectorDef).type}`);
      }
    }
  });
}

File: packages/python/corti_agent_sdk/connectors.py
from __future__ import annotations

import re
from typing import Any, Dict, List, Optional

from .types import A2aConnector, ConnectorDef, CortiAgentConnector, McpConnector, RegistryConnector


# ── Public connector factories ───────────────────────────────────────────────

class connectors:
    """
    Factory helpers for building connector definitions.

    Usage::

        connectors.from_agent(agent_id="abc-123")
        connectors.mcp(mcp_url="https://mcp.corti.ai")
        connectors.registry(name="@corti/medical-coding")
    """

    @staticmethod
    def from_agent(agent_id: str) -> CortiAgentConnector:
        """Reference another Corti agent as a sub-agent connector."""
        return {"type": "cortiAgent", "agent_id": agent_id}

    @staticmethod
    def mcp(
        mcp_url: str,
        *,
        name: Optional[str] = None,
        transport: Optional[str] = None,
        token: Optional[str] = None,
    ) -> McpConnector:
        """Attach an MCP server directly to the agent."""
        c: McpConnector = {"type": "mcp", "mcp_url": mcp_url}
        if name is not None:
            c["name"] = name
        if transport is not None:
            c["transport"] = transport  # type: ignore[typeddict-item]
        if token is not None:
            c["token"] = token
        return c

    @staticmethod
    def registry(name: str, *, system_prompt: Optional[str] = None) -> RegistryConnector:
        """Reference a named expert from the Corti registry."""
        c: RegistryConnector = {"type": "registry", "name": name}
        if system_prompt is not None:
            c["system_prompt"] = system_prompt
        return c

    @staticmethod
    def a2a(a2a_url: str) -> A2aConnector:
        """A2A connector (reserved – not yet supported)."""
        return {"type": "a2a", "a2a_url": a2a_url}


# ── Internal: ConnectorDef[] → experts[] (REST API shape) ───────────────────

def _mcp_url_to_name(url: str) -> str:
    m = re.match(r"^https?://([^/?#]+)", url)
    hostname = m.group(1) if m else ""
    return re.sub(r"[^a-z0-9-]", "", hostname.replace(".", "-"))[:48] or "mcp-server"


def connectors_to_experts(defs: List[ConnectorDef]) -> List[Dict[str, Any]]:
    """Translate wrapper ConnectorDef list → REST API experts array."""
    experts: List[Dict[str, Any]] = []

    for conn in defs:
        t = conn["type"]

        if t == "mcp":
            mcp = conn  # type: ignore[assignment]
            name = mcp.get("name") or _mcp_url_to_name(mcp["mcp_url"])
            server: Dict[str, Any] = {
                "name": name,
                "transportType": mcp.get("transport", "sse"),
                "authorizationType": "bearer" if mcp.get("token") else "inherit",
                "url": mcp["mcp_url"],
            }
            if mcp.get("token"):
                server["token"] = mcp["token"]
            experts.append(
                {
                    "type": "new",
                    "name": name,
                    "description": f"MCP server at {mcp['mcp_url']}",
                    "mcpServers": [server],
                }
            )

        elif t == "registry":
            reg = conn  # type: ignore[assignment]
            expert: Dict[str, Any] = {"type": "reference", "name": reg["name"]}
            if reg.get("system_prompt"):
                expert["systemPrompt"] = reg["system_prompt"]
            experts.append(expert)

        elif t == "cortiAgent":
            ca = conn  # type: ignore[assignment]
            experts.append({"type": "reference", "id": ca["agent_id"]})

        elif t == "a2a":
            a2a = conn  # type: ignore[assignment]
            raise ValueError(
                f"[AgentSDK] A2A connectors are not yet supported (url: {a2a['a2a_url']}). "
                'Use type "mcp" with an MCP-compatible endpoint instead.'
            )

        else:
            raise ValueError(f"[AgentSDK] Unknown connector type: {t!r}")

    return experts
